### PR TITLE
Support async NIO requests via reactive-streams

### PR DESCRIPTION
**Pull request by johnmcclean-aol**
Impl for #340

### CODE REVIEW

File: micro-grizzly-with-jersey/src/test/java/app/publisher/binder/direct/AsyncPublisherTest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        server.stop();
+    }
+
+    @Test
+    public void runAppAndBasicTest() throws InterruptedException, ExecutionException {
+
+
+        //DirectMyIncovationHandler.captured=false;
+        assertThat(rest.get("http://localhost:8080/binder/test"), is("hello world!"));
+        //assertTrue(DirectMyIncovationHandler.captured);
```
Comment (morrowgi) [inline micro-grizzly-with-jersey/src/test/java/app/publisher/binder/direct/AsyncPublisherTest.java]: Can we remove the commented out lines?

File: micro-grizzly-with-jersey/src/test/java/app/custom/binder/direct/CustomBinder4.java [hunk 2]
```diff
@@ -11,5 +11,7 @@ protected void configure() {
         // this is where the magic happens!
         bind(DirectCustomResourceInvocationHandlerProvider.class).to(
                 ResourceMethodInvocationHandlerProvider.class);
+
```
Comment (tonyfinn) [inline micro-grizzly-with-jersey/src/test/java/app/custom/binder/direct/CustomBinder4.java]: Remove extra whitespace

File: micro-grizzly-with-jersey/src/test/java/app/publisher/binder/direct/AsyncPublisherTest.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @After
+    public void stopServer() {
+        server.stop();
+    }
+
+    @Test
+    public void runAppAndBasicTest() throws InterruptedException, ExecutionException {
+
```
Comment (tonyfinn) [inline micro-grizzly-with-jersey/src/test/java/app/publisher/binder/direct/AsyncPublisherTest.java]: Too much whitespace here also.

File: micro-jersey/src/main/java/com/aol/micro/server/rest/jersey/AsyncDispatcher.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    public AsyncDispatcher(ResourceMethodDispatcher originalDispatcher) {
+        this.originalDispatcher = originalDispatcher;
+    }
+
+    @AllArgsConstructor
+    @NoArgsConstructor
+   static class AsyncDispatcherProvider implements Provider{
+        @Context
+        private ServiceLocator serviceLocator;
```
Comment (tonyfinn) [inline micro-jersey/src/main/java/com/aol/micro/server/rest/jersey/AsyncDispatcher.java]: Formatting here too

File: micro-jersey/src/main/java/com/aol/micro/server/rest/jersey/AsyncInvocationHandler.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+
+public class AsyncInvocationHandler implements ResourceMethodInvocationHandlerProvider {
+
+
+    @Override
+    public InvocationHandler create(Invocable invocable) {
+        Class returnType = invocable.getRawResponseType();
```
Comment (tonyfinn) [inline micro-jersey/src/main/java/com/aol/micro/server/rest/jersey/AsyncInvocationHandler.java]: This is never used?

File: readme.md [hunk 6]
```diff
@@ -46,6 +46,34 @@ See the response  *hello world!*
...
 
+## Easy to use async NIO based REST
+
+Return any reactive-streams Publisher from your REST end point to make them execute asynchronously automatically.
+
+E.g. Using Future from [cyclops-react](cyclops-react.io)
+```java
+   @GET
+   public Future<String> myEndPoint(){
+	  return Future.ofSupplier(()->{
```
Comment (tonyfinn) [inline readme.md]: Formatting in docs

File: micro-grizzly-with-jersey/src/test/java/app/publisher/binder/direct/AsyncResource.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    public Future<String> myEndPoint() {
+        return Future.ofSupplier(() -> {
+            sleep();
+            return "hello world!";
+        }, Executors.newFixedThreadPool(1));
+    }
+
+    @GET
+    public ReactiveSeq<String> async2() {
+        return Spouts.publishOn(Stream.of("hello"), Executors.newFixedThreadPool(1));
```
Comment (tonyfinn) [inline micro-grizzly-with-jersey/src/test/java/app/publisher/binder/direct/AsyncResource.java]: Is this used?

File: micro-jersey/src/main/java/com/aol/micro/server/rest/jersey/AsyncInvocationHandler.java (hunk 5 above)
Comment (johnmcclean-aol) [inline micro-jersey/src/main/java/com/aol/micro/server/rest/jersey/AsyncInvocationHandler.java]: Good point, I think this class may not be needed.

File: micro-grizzly-with-jersey/src/test/java/app/publisher/binder/direct/AsyncResource.java (hunk 7 above)
Comment (johnmcclean-aol) [inline micro-grizzly-with-jersey/src/test/java/app/publisher/binder/direct/AsyncResource.java]: It should be, I think only during manual testing. Will add an automated test for it.

**Pull request merged**